### PR TITLE
standardize post-call treatment of get_dst_info, delay sorting calls

### PR DESCRIPTION
**Pull request by jbrockmendel**
@jreback we discussed how there are a bunch of functions that do really similar things with `get_dst_info` but have their slight idiosyncrasies.  This standardizes them, and is really verbose so as to delay certain calls until absolutely necessary.

After this we can see about de-duplicating the 6ish occurrences of really similar code by passing a function pointer or something.

### CODE REVIEW

File: pandas/_libs/tslibs/timezones.pyx [hunk 1]
```diff
@@ -258,12 +258,17 @@ cdef object get_dst_info(object tz):
...
-                trans = np.array([], dtype='M8[ns]')
-                deltas = np.array([], dtype='i8')
-                typ = None
+                # 2018-07-12 this is not reached in the tests, and this case
+                # is not handled in any of the functions that call
+                # get_dst_info.  If this case _were_ hit the calling
+                # functions would then hit an IndexError because they assume
+                # `deltas` is non-empty.
+                # (under the just-deleted code that returned empty arrays)
+                raise ValueError(tz)
```
Comment (jreback) [inline pandas/_libs/tslibs/timezones.pyx]: maybe an AssertionError for paths that are not hit (or can just use an assert)

**Comment by jreback** [comment]
looks fine. is there any perf diff?

**Comment by jbrockmendel** [comment]
Noise, but at least good noise I guess.

```
asv continuous -f 1.1 -E virtualenv master HEAD -b timeseries
[...]
       before           after         ratio
     [3cb64bd2]       [76e0c080]
-     3.92±0.03μs      3.55±0.01μs     0.91  timeseries.DatetimeIndex.time_get('dst')
-       136±0.1ms       120±0.06ms     0.89  timeseries.DatetimeIndex.time_to_pydatetime('tz_aware')
-     29.8±0.07ms       26.0±0.1ms     0.87  timeseries.Iteration.time_iter_preexit(<function period_range at 0x7f36ac823de8>)
-         935±1μs          731±2μs     0.78  timeseries.DatetimeIndex.time_to_pydatetime('dst')
```

```
taskset 6 asv continuous -f 1.1 -E virtualenv master HEAD -b timeseries
[...]
       before           after         ratio
     [3cb64bd2]       [76e0c080]
-       294±0.2ms          263±1ms     0.89  timeseries.ToDatetimeISO8601.time_iso8601_tz_spaceformat
-     7.38±0.03μs      6.50±0.04μs     0.88  timeseries.DatetimeIndex.time_get('tz_aware')
-      10.6±0.1ms         7.75±0ms     0.73  timeseries.ToDatetimeISO8601.time_iso8601
```

**Comment by jreback** [comment]
thanks @jbrockmendel happy to have cleanups

**Comment by jbrockmendel** [comment]
> warning: pandas/_libs/tslibs/conversion.pyx:614:39: Buffer unpacking not optimized away

I'm pretty sure this has to do with the return type being `int64_t[:]`, will see if there's anything that can be done about it.

**Pull request merged**